File: typescript/graph/dijkstra.ts
import Graph from "./graph.ts";
import PriorityQueue from "./priority_queue.ts";

const dijkstra = (g: Graph, start: number) => {
  const n = g.getVerticesNum();

  const pred: number[] = Array(n).fill(-1);
  const dist: number[] = Array(n).fill(Number.MAX_SAFE_INTEGER);

  dist[start] = 0;
  const pq = new PriorityQueue<number>();

  for (let u = 0; u < n; u++) {
    const node = { value: u, priority: dist[u] };
    pq.enqueue(node);
  }

  while (!pq.isEmpty()) {
    const u = pq.dequeue()?.value as number;

    for (const [v, w] of g.getVertices(u)) {
      const newLength = dist[u] + w;

      if (newLength < dist[v]) {
        pq.updatePriority({ value: v, priority: newLength });
        dist[v] = newLength;
        pred[v] = u;
      }
    }
  }

  return { pred, dist };
};

export default dijkstra;

File: dart/lib/graph/dijkstra.dart
import 'package:algorithm/collections/priority_queue.dart';
import 'package:algorithm/graph/graph.dart';

const int64Max = 92233720368547758;

List<List<int>> dijkstra(Graph g, int start) {
  final n = g.n;

  final pred = List.generate(n, (_) => -1);
  final dist = List.generate(n, (_) => int64Max);

  dist[start] = 0;
  final pq = PriorityQueue<int>();

  for (var u = 0; u < n; u++) {
    final node = Node(u, dist[u]);
    pq.enqueue(node);
  }

  while (!pq.isEmpty()) {
    final u = pq.dequeue()?.value as int;

    for (final pair in g.getVertices(u)) {
      final int v = pair.first;
      final int w = pair.second;
      final newLength = dist[u] + w;

      if (newLength < dist[v]) {
        pq.updatePriority(Node(v, newLength));
        dist[v] = newLength;
        pred[v] = u;
      }
    }
  }

  return [pred, dist];
}
